js/source/source.js

```diff
@@ -160,34 +114,28 @@ exports._querySourceFeatures = function(params) {
  * map.addSource('some id', sourceObj); // add
  * map.removeSource('some id');  // remove
  */
-exports.create = function(source) {
-    // This is not at file scope in order to avoid a circular require.
-    var sources = {
-        vector: require('./vector_tile_source'),
-        raster: require('./raster_tile_source'),
-        geojson: require('./geojson_source'),
-        video: require('./video_source'),
-        image: require('./image_source')
-    };
-
-    return exports.is(source) ? source : new sources[source.type](source);
+exports.create = function(id, source, dispatcher, onChange, callback) {
+    return exports.is(source) ? source :
+        sourceTypes[source.type](id, source, dispatcher, onChange, callback);
 };
 
 exports.is = function(source) {
-    // This is not at file scope in order to avoid a circular require.
-    var sources = {
-        vector: require('./vector_tile_source'),
-        raster: require('./raster_tile_source'),
-        geojson: require('./geojson_source'),
-        video: require('./video_source'),
-        image: require('./image_source')
-    };
+    // TODO: with the Source interface being factory-based, instanceof doesn't
+    // work as naturally for this.  Is Source.is() still needed?
+
+    return source && source.id;
+    // for (var type in sourceTypes) {
+    //     if (source instanceof sourceTypes[type]) {
+    //         return true;
+    //     }
+    // }
+    // return false;
+};
```

Comment:
If using a `new`-based API is going to be cleaner than using a factory-based API, I'm all for it.